### ISSUE TITLE
[propasal] Add unit tests

### ISSUE THREAD

**Issue by tymondesigns** [inuitcss/inuitcss#133]
So, on a scale of "Hell No" to "Hell yeah" how interested is everybody in adding some unit tests?

I would propose adding [this package](https://github.com/oddbird/true) to handle the testing of the functions and mixins.

for example:

```sass
@include test('Font Size [mixin]') {
  @include assert('Outputs a font size and line height based on a given value.') {
    @include output {
      @include font-size(16px);
    }

    @include expect {
      font-size: 16px;
      font-size: 1rem;
      line-height: 1.5;
    }
  }
}
```

Again I'd be happy to get the ball rolling on this...

**Comment by anselmh** [comment]
Hey, thanks for the suggestion. I’m a bit split on this topic. Usually I think testing code is a desirable goal but testing CSS doesn’t work very well and needs crazy stuff to make it work (the code to write tests, the lib that you suggested) and it requires a pretty high level of knowledge on this unit testing if you want to contribute code.

Therefore I’m currently voting not to do it here, but this is something we should discuss / agree on as a whole team. 
So, @inuitcss/core please vote 👎 or 👍 here on my post. Thanks!

**Comment by nenadjelovac** [comment]
@tymondesigns thanks for the suggestion! However, I have to vote 👎  because of the reasons @anselmh said, and because of the requirement that [simplicity is paramount](https://github.com/inuitcss/inuitcss/blob/develop/CONTRIBUTING.md#simplicity-is-paramount). Hope it makes sense.

**Comment by csswizardry** [comment]
Holy s—t, this is insane: I was literally just thinking about adding tests to inuitcss! I wasn’t sure how best to do it (external lib or writing something small to cover our specific mixins) but let me take a look at true.

**Comment by csswizardry** [comment]
@nenadjelovac: I don’t think this encroaches on simplicity: tests will allow us to refactor any mixins etc. and be immediately notified about breakages. They would likely be private to the framework.

**Comment by tymondesigns** [comment]
@csswizardry My thoughts exactly. The external end-user experience will be unaffected, and this will simply serve as an internal tool to provide the confidence when refactoring code.

**Comment by nenadjelovac** [comment]
@csswizardry 
>They would likely be private to the framework.

if that is the case i'd probably change my vote, but I'd have to see how it's gonna get implemented ;)

**Comment by tymondesigns** [comment]
If no-one beats me to it, I can knock a little POC together this evening when I'm home...

**Comment by anselmh** [comment]
So we’d only test internal functions and mixins, right? This would be fine for me. Still not sure if it’s a good idea but if you all agree, then let’s do this.

**Comment by dnixx** [comment]
Tests don't add complexity to userland code. It can even be a dev dependency :)

**Comment by csshugs** [comment]
Just want to throw in the question: Is inuitcss' Sass stuff really that complex to justify adding the complexity of unit tests? Becuase although it's not complexity for inuitcss' users, it *is* complexity for us (read: the inuitcss dev team).

I mean, this is not a huge Sass library a la [Bourbon](http://bourbon.io/) with much complex stuff in it. The most complex Sass code in inuitcss at the moment is probably [this](https://github.com/inuitcss/inuitcss/blob/develop/utilities/_utilities.widths.scss).

I'm just sharing my opinion here which is: I wouldn't need Sass unit tests in inuitcss. 

If you guys consider it being a good idea though, I'm on board.

**Comment by tymondesigns** [comment]
I have created the PR #134 so that we can see what it might look like.

here's a screenshot of the terminal for reference.

![image](https://cloud.githubusercontent.com/assets/1801923/17492765/36362632-5da5-11e6-951f-1ab9d371dbbe.png)

@csshugs I can understand your point of view, and I somewhat agree, but for me at least I don't see it as being that complicated to write the tests, versus the benefits they would yield.

@csswizardry Do you have any more thoughts on this now that there's some code to look at?

**Comment by florianbouvot** [comment]
Like @csshugs and @anselmh I'm not sure it's necessary for this kind of project.

**Comment by anselmh** [comment]
So, my initial feedback was based on the idea of assuming that all code of inuitcss will be unit tested. This is something I’d strongly object, since it makes contributions too hard and would also require a lot of additional work by all core contributors not only initially but also over-time.

However, testing our internal mixins and functions only, makes totally sense to me in a sense that we have easy unit test results and a failure indicator when we refactor something or add a new feature. I don’t think it’d hurt to add these few tests here, to be honest.

**Comment by csswizardry** [comment]
Tests (in general) are a great idea IMO. It gives us programmatic safety. I’m stuck in a coffee shop in the middle of Buenos Aires with 10% battery on my laptop right now, so I can’t take a close look, but thanks so much, @tymondesigns! I’ll delve into this ASAP :)

**Comment by tymondesigns** [comment]
@csswizardry no problem! 😄

**Comment by nenadjelovac** [comment]
I thought the same as @anselmh! But now I get it. 

While on the topic, shouldn't we lint the code as well?

**Comment by anselmh** [comment]
@nenadjelovac different, topic so please create a new issue for it. In short: Yes, we should, if possible – preferrably by using http://stylelint.io/ :)

**Comment by anselmh** [comment]
This landed in 6ce1a81ec501a6f3c70059d028336ac77234c107.
